File: platforms/ios/Sources/Ratex/RaTeXRenderer.swift
// RaTeXRenderer.swift — CoreGraphics + CoreText renderer for a RaTeX DisplayList.
//
// Key insight: GlyphPath.commands are PLACEHOLDER RECTANGLES (bounding boxes),
// not real glyph outlines. Glyphs must be drawn with CTFont using font + char_code.
// Non-glyph items (Line, Rect, Path) use their commands directly.
//
// Fonts are registered automatically on first use via RaTeXFontLoader.ensureLoaded(); or call loadFromBundle() at startup.

import CoreGraphics
import CoreText
import Foundation

public struct RaTeXRenderer {
    public let displayList: DisplayList
    /// Font size in points. All em-unit coordinates are multiplied by this value.
    public let fontSize: CGFloat

    public init(displayList: DisplayList, fontSize: CGFloat = 24) {
        self.displayList = displayList
        self.fontSize = fontSize
    }

    // MARK: - Dimensions (in points)

    public var width:       CGFloat { CGFloat(displayList.width)  * fontSize }
    public var height:      CGFloat { CGFloat(displayList.height) * fontSize }
    public var depth:       CGFloat { CGFloat(displayList.depth)  * fontSize }
    public var totalHeight: CGFloat { height + depth }

    // MARK: - Drawing

    /// Draw the formula into `context`.
    /// The origin is the top-left of the formula's bounding box.
    public func draw(in context: CGContext) {
        for item in displayList.items {
            switch item {
            case .glyphPath(let g): drawGlyph(g, in: context)
            case .line(let l):      drawLine(l, in: context)
            case .rect(let r):      drawRect(r, in: context)
            case .path(let p):      drawPath(p, in: context)
            case .unknown:
                // Forward compatibility: ignore items newer than this decoder.
                break
            }
        }
    }

    // MARK: - Private helpers

    /// Convert em units to points.
    private func pt(_ em: Double) -> CGFloat { CGFloat(em) * fontSize }

    private func cgColor(_ c: RaTeXColor) -> CGColor {
        CGColor(red: CGFloat(c.r), green: CGFloat(c.g),
                blue: CGFloat(c.b), alpha: CGFloat(c.a))
    }

    // MARK: Glyph (CoreText)

    /// Map RaTeX font ID ("Math-Italic") to the KaTeX PostScript name ("KaTeX_Math-Italic").
    private func postScriptName(for fontId: String) -> String {
        "KaTeX_\(fontId)"
    }

    private func drawGlyph(_ g: GlyphPathData, in ctx: CGContext) {
        // GlyphPath.commands are placeholder rects — ignore them.
        // Draw the actual character using CoreText.
        guard let scalar = Unicode.Scalar(g.charCode) else { return }
        let char = String(Character(scalar))

        let psName  = postScriptName(for: g.font)
        let ctFont  = CTFontCreateWithName(psName as CFString, pt(g.scale), nil)
        let color   = cgColor(g.color)

        let attrs: [CFString: Any] = [
            kCTFontAttributeName:                    ctFont,
            kCTForegroundColorAttributeName:         color,
        ]
        let attrStr = CFAttributedStringCreate(nil, char as CFString, attrs as CFDictionary)!
        let line    = CTLineCreateWithAttributedString(attrStr)

        ctx.saveGState()

        // In UIKit's draw(_:) context, Y increases downward (UIKit has already flipped the CTM).
        // The GlyphPath y-coordinate is the alphabetic baseline measured downward from the top.
        ctx.translateBy(x: pt(g.x), y: pt(g.y))

        // CoreText draws upward from the baseline.
        // UIKit's flipped CTM would make CoreText text appear upside-down.
        // textMatrix = (1,0,0,-1) counteracts the y-flip for text rendering.
        ctx.textMatrix = CGAffineTransform(a: 1, b: 0, c: 0, d: -1, tx: 0, ty: 0)

        CTLineDraw(line, ctx)

        ctx.restoreGState()
    }

    // MARK: Line

    private func drawLine(_ l: LineData, in ctx: CGContext) {
        ctx.saveGState()
        let t = max(0.5, pt(l.thickness))
        let halfT = t / 2
        if l.dashed {
            ctx.setStrokeColor(cgColor(l.color))
            ctx.setLineWidth(t)
            ctx.setLineCap(.butt)
            let dashLen = t * 3
            ctx.setLineDash(phase: 0, lengths: [dashLen, dashLen])
            ctx.move(to: CGPoint(x: pt(l.x), y: pt(l.y)))
            ctx.addLine(to: CGPoint(x: pt(l.x) + pt(l.width), y: pt(l.y)))
            ctx.strokePath()
        } else {
            ctx.setFillColor(cgColor(l.color))
            ctx.fill(CGRect(x: pt(l.x), y: pt(l.y) - halfT,
                            width: pt(l.width), height: t))
        }
        ctx.restoreGState()
    }

    // MARK: Rect

    private func drawRect(_ r: RectData, in ctx: CGContext) {
        ctx.saveGState()
        ctx.setFillColor(cgColor(r.color))
        ctx.fill(CGRect(x: pt(r.x), y: pt(r.y),
                        width: pt(r.width), height: pt(r.height)))
        ctx.restoreGState()
    }

    // MARK: Path (radical signs, delimiters, etc.)

    private func makeCGPath(from commands: [PathCommand], dx: Double = 0, dy: Double = 0) -> CGPath {
        let path = CGMutablePath()
        let ox = pt(dx), oy = pt(dy)
        for cmd in commands {
            switch cmd {
            case .moveTo(let x, let y):
                path.move(to: CGPoint(x: ox + pt(x), y: oy + pt(y)))
            case .lineTo(let x, let y):
                path.addLine(to: CGPoint(x: ox + pt(x), y: oy + pt(y)))
            case .cubicTo(let x1, let y1, let x2, let y2, let x, let y):
                path.addCurve(to:       CGPoint(x: ox + pt(x),  y: oy + pt(y)),
                              control1: CGPoint(x: ox + pt(x1), y: oy + pt(y1)),
                              control2: CGPoint(x: ox + pt(x2), y: oy + pt(y2)))
            case .quadTo(let x1, let y1, let x, let y):
                path.addQuadCurve(to:      CGPoint(x: ox + pt(x),  y: oy + pt(y)),
                                  control: CGPoint(x: ox + pt(x1), y: oy + pt(y1)))
            case .close:
                path.closeSubpath()
            case .unknown:
                // Forward compatibility: ignore newer command variants.
                break
            }
        }
        return path
    }

    private func drawPath(_ p: PathData, in ctx: CGContext) {
        ctx.saveGState()
        let cgPath = makeCGPath(from: p.commands, dx: p.x, dy: p.y)
        ctx.addPath(cgPath)
        let color = cgColor(p.color)
        if p.fill {
            ctx.setFillColor(color)
            ctx.fillPath()
        } else {
            ctx.setStrokeColor(color)
            ctx.strokePath()
        }
        ctx.restoreGState()
    }
}

File: platforms/android/src/main/kotlin/io/ratex/RaTeXRenderer.kt
// RaTeXRenderer.kt — Android Canvas renderer for a RaTeX DisplayList.

package io.ratex

import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path as AndroidPath
import android.graphics.Typeface
import android.util.Log

/**
 * Renders a [DisplayList] onto an Android [Canvas].
 *
 * All em-unit coordinates are multiplied by [fontSize] (sp/px) to get screen coordinates.
 *
 * GlyphPath items are **placeholder rectangles** from Rust; glyphs are drawn with
 * KaTeX fonts. [RaTeXView] calls [RaTeXFontLoader.ensureLoaded] on first use. Pass
 * [typefaceLoader] (e.g. [RaTeXFontLoader.getTypeface]). If null, glyphs are drawn as black rectangles.
 *
 * @param displayList   The layout output from [RaTeXEngine.parse].
 * @param fontSize      Font size in pixels. Matches the desired display size on screen.
 * @param typefaceLoader Optional (fontId: String) -> Typeface? for glyph rendering.
 */
class RaTeXRenderer(
    val displayList: DisplayList,
    val fontSize: Float,
    private val typefaceLoader: ((String) -> Typeface?)? = null,
) {
    // MARK: - Dimensions in pixels

    val widthPx:       Float get() = (displayList.width  * fontSize).toFloat()
    val heightPx:      Float get() = (displayList.height * fontSize).toFloat()
    val depthPx:       Float get() = (displayList.depth  * fontSize).toFloat()
    val totalHeightPx: Float get() = heightPx + depthPx

    // MARK: - Drawing

    /** Draw the formula into [canvas]. The canvas origin is the top-left of the bounding box. */
    fun draw(canvas: Canvas) {
        for (item in displayList.items) {
            when (item) {
                is DisplayItem.GlyphPath -> drawGlyph(canvas, item)
                is DisplayItem.Line      -> drawLine(canvas, item)
                is DisplayItem.Rect      -> drawRect(canvas, item)
                is DisplayItem.Path      -> drawPath(canvas, item)
            }
        }
    }

    // MARK: - Private helpers

    private fun Float.em() = this * fontSize
    private fun Double.em() = (this * fontSize).toFloat()

    private fun Paint.applyColor(c: RaTeXColor) { color = c.toArgb() }

    private fun buildAndroidPath(commands: List<PathCommand>, dx: Float = 0f, dy: Float = 0f): AndroidPath {
        val path = AndroidPath()
        for (cmd in commands) {
            when (cmd) {
                is PathCommand.MoveTo  -> path.moveTo(dx + cmd.x.em(), dy + cmd.y.em())
                is PathCommand.LineTo  -> path.lineTo(dx + cmd.x.em(), dy + cmd.y.em())
                is PathCommand.CubicTo -> path.cubicTo(
                    dx + cmd.x1.em(), dy + cmd.y1.em(),
                    dx + cmd.x2.em(), dy + cmd.y2.em(),
                    dx + cmd.x.em(),  dy + cmd.y.em())
                is PathCommand.QuadTo  -> path.quadTo(
                    dx + cmd.x1.em(), dy + cmd.y1.em(),
                    dx + cmd.x.em(),  dy + cmd.y.em())
                PathCommand.Close      -> path.close()
            }
        }
        return path
    }

    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val textPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { isAntiAlias = true }

    private fun drawGlyph(canvas: Canvas, g: DisplayItem.GlyphPath) {
        val typeface = typefaceLoader?.invoke(g.font) ?: return
        val codePoint = g.charCode.toInt()
        if (!Character.isValidCodePoint(codePoint)) {
            Log.w("RaTeXRenderer", "Invalid Unicode code point: $codePoint (0x${codePoint.toString(16)}) in font '${g.font}' — skipping glyph")
            return
        }
        canvas.save()
        canvas.translate(g.x.em(), g.y.em())
        val str = String(Character.toChars(codePoint))
        textPaint.typeface = typeface
        textPaint.textSize = fontSize * g.scale.toFloat()
        textPaint.color = g.color.toArgb()
        textPaint.style = Paint.Style.FILL
        canvas.drawText(str, 0f, 0f, textPaint)
        canvas.restore()
    }

    private fun drawLine(canvas: Canvas, l: DisplayItem.Line) {
        val t = maxOf(0.5f, (l.thickness * fontSize).toFloat())
        val halfT = t / 2f
        val left   = l.x.em()
        val top    = l.y.em() - halfT
        val right  = (l.x + l.width).em()
        val bottom = l.y.em() + halfT
        paint.applyColor(l.color)
        if (l.dashed) {
            val dashLen = t * 3f
            paint.style = Paint.Style.STROKE
            paint.strokeWidth = t
            paint.strokeCap = Paint.Cap.BUTT
            paint.pathEffect = android.graphics.DashPathEffect(floatArrayOf(dashLen, dashLen), 0f)
            val path = android.graphics.Path()
            path.moveTo(left, l.y.em())
            path.lineTo(right, l.y.em())
            canvas.drawPath(path, paint)
            paint.pathEffect = null
        } else {
            paint.style = Paint.Style.FILL
            canvas.drawRect(left, top, right, bottom, paint)
        }
    }

    private fun drawRect(canvas: Canvas, r: DisplayItem.Rect) {
        paint.style = Paint.Style.FILL
        paint.applyColor(r.color)
        canvas.drawRect(
            r.x.em(), r.y.em(),
            (r.x + r.width).em(), (r.y + r.height).em(),
            paint)
    }

    private fun drawPath(canvas: Canvas, p: DisplayItem.Path) {
        val path = buildAndroidPath(p.commands, p.x.em(), p.y.em())
        paint.applyColor(p.color)
        paint.style = if (p.fill) Paint.Style.FILL else Paint.Style.STROKE
        canvas.drawPath(path, paint)
    }
}
